feat: add support for Android user-installed certificates

jfly opened the pull request:
## Description

This is part of https://github.com/immich-app/immich/issues/15230.

Frustratingly, Dart/Flutter ignores user-installed certificates. Working around this requires rooting your Android device to install certificates as "system" certs, which isn't an option for everyone.

This is a known issue with Dart, see
https://github.com/dart-lang/sdk/issues/50435 and
https://github.com/flutter/flutter/issues/56607 for details.

I have read through
<https://github.com/immich-app/immich/issues/15230> and <https://github.com/immich-app/immich/discussions/13555>, and I understand that switching to
[`cronnet_http`](https://github.com/immich-app/immich/pull/14335#issuecomment-2695096512) would also resolve this. While that may be the correct long-term approach, it looks like there are [a lot of network codepaths](https://github.com/immich-app/immich/issues/15230#issuecomment-2584057446) in Immich, and as I know basically nothing about Dart, nor Flutter, nor Immich's codebase, I thought this would be a better short term approach.

This depends on my fork of
`johnstef99/flutter_user_certificates_android`, which I've sent a PR for here
<https://github.com/johnstef99/flutter_user_certificates_android/pull/2>. If y'all don't like the supply chain implications of that, I'm happy to inline the implementation here instead.

## How Has This Been Tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->
I have step-ca and DNS configured so that immich.mm resolves to an immich server with a certificate issued by my step-ca certificate authority. I've installed that certificate authority as a user-installed certificate on my Android phone.

- [x] With a fresh install of immich, I can configure https://immich.mm as a server
- [x] After logging in, I can view photos, upload photos and video, and play video.

<details><summary><h2>Screenshots (if appropriate)</h2></summary>

![2025-07-08_18-06-21_pattern](https://github.com/user-attachments/assets/47df04c4-c607-441d-be0c-581545242b1b)

The help text on this option is no longer true. I'd suggest that we deprecate it in favor of encouraging people to add certificates to their OS:

![2025-07-08_17-59-03_pattern](https://github.com/user-attachments/assets/3e9ac967-8a74-48eb-b408-8ed235d334b1)

</details>

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation if applicable
- [x] I have no unrelated changes in the PR.
- [ ] I have confirmed that any new dependencies are strictly necessary. **Strictly? No. See my offer above about to inline the implementation if that's preferred.**
- [ ] I have written tests for new code (if applicable)
- [ ] I have followed naming conventions/patterns in the surrounding code
- [ ] All code in `src/services/` uses repositories implementations for database calls, filesystem operations, etc.
- [ ] All code in `src/repositories/` is pretty basic/simple and does not have any immich specific logic (that belongs in `src/services/`)

jfly committed:
feat: add support for Android user-installed certificates

This is part of https://github.com/immich-app/immich/issues/15230.

Frustratingly, Dart/Flutter ignores user-installed certificates. Working
around this requires rooting your Android device to install certificates
as "system" certs, which isn't an option for everyone.

This is a known issue with Dart, see
https://github.com/dart-lang/sdk/issues/50435 and
https://github.com/flutter/flutter/issues/56607 for details.

I have read through
<https://github.com/immich-app/immich/issues/15230> and
<https://github.com/immich-app/immich/discussions/13555>, and I
understand that switching to
[`cronnet_http`](https://github.com/immich-app/immich/pull/14335#issuecomment-2695096512)
would also resolve this. While that may be the correct long-term
approach, it looks like there are [a lot of network
codepaths](https://github.com/immich-app/immich/issues/15230#issuecomment-2584057446)
in Immich, and as I know basically nothing about Dart, nor Flutter, nor
Immich's codebase, I thought this would be a better short term approach.

For the record: the important code here is copied from my fork of
`johnstef99/flutter_user_certificates_android`. See this PR for more
context: <https://github.com/johnstef99/flutter_user_certificates_android/pull/2>.

jfly commented on mobile/pubspec.yaml:
```diff
@@ -36,6 +36,13 @@ dependencies:
+  # TODO: switch to regular `flutter_user_certificates_android` when this PR is
+  # merged:
+  # <https://github.com/johnstef99/flutter_user_certificates_android/pull/2>.
+  flutter_user_certificates_android:
+    git:
+      url: https://github.com/jfly/flutter_user_certificates_android.git
+      ref: "add-trusting"
```
I totally understand. Happy to do any of the following:

1. Pin to a specific commit, as you mention.
2. Depend on the unpatched `flutter_user_certificates_android` and vendor just the changes from https://github.com/johnstef99/flutter_user_certificates_android/pull/2.
3. Vendor the whole thing. We're not talking about a lot of code here.

Just let me know which approach to take.

jfly replied:
@bo0tzz, friendly ping here. Which approach would you like me to take?

jfly commented on mobile/android/app/src/main/res/xml/network_security_config.xml:
```diff
@@ -0,0 +1,9 @@
+  <base-config>
+    <trust-anchors>
+      <certificates src="system"/>
+      <certificates src="user"/>
```
I don't know, but everything I know about android development I learned in the last week while researching this issue.

shenlong-tanwen commented:
@jfly Hey, Thanks a lot for working on this. However, if we are to merge this in, rather than deprecating the existing "Allow Self signed certificates" toggle, we could just remove all the custom verification logic all together. We don't need a lot of different codepath to address the same issue. However, this would mean that we would also need to test this exhaustively to make sure it does not break anything from the current setup. So things like:
- mTLS
- Self-signed certificates
- Custom CAs

For all cases, we'd need to test both the dart side of things and the native side of things. Most network calls are from the dart side. Video player, Asset download are made from native side and we should validate these as well. If we are to remove the self signed toggle, we've to make sure everything works on iOS as well.

You could just inline all the changes. We are migrating to using pigeon for all message passing between flutter and the platform side, so you can add an API to fetch the certificates and load it in the dart's SecurityContext. We might also need some documentation change to inform users about installing certificates to the OS cert store, rather than loading it in the app like they used to.

The core team currently has got it's hands full with working towards the stable release. I know this is a significant undertaking for your free time, but if you can get this properly tested after removing the existing toggle and adding in your changes, we can happily merge this in. If you need help setting up specific test scenarios or access to different certificate setups, let us know - we can coordinate with the community to help test various configurations.

jfly commented:
> I know this is a significant undertaking for your free time, but if you can get this properly tested after removing the existing toggle and adding in your changes, we can happily merge this in.

Sure, I'm happy to take this on. I've converted my understanding of your requirements into a checklist below. Please let me know if I'm missing anything:

- [ ] Inline the changes (drop dependency on `johnstef99/flutter_user_certificates_android`)
  - [ ] Use pigeon for message passing between dart and kotlin.
- [ ] Remove "Allow self-signed SSL certificates" checkbox
- [ ] Docs? @shenlong-tanwen, please advise. IMO this is unnecessary, it's just fixing a bug and making things work the way users would expect. I do think a release note would be useful, is there a good place to add that?
- [ ] Test on Android. Confirm you can view photos, upload photos and video, play video, and download assets.
  - [ ] mTLS (aka "SSL Client Certificate")
  - [ ] Self-signed certificates
  - [ ] Custom CAs
- [ ] Test on iOS. Confirm you can view photos, upload photos and video, play video, and download assets. @shenlong-tanwen, I will need help with this as I do not have access to an iOS device, nor development machine.
  - [ ] mTLS (aka "SSL Client Certificate")
  - [ ] Self-signed certificates
  - [ ] Custom CAs

richardlt commented:
Personally I agree with your second point, since this feature here will not strictly replace the existing flag it could be a breaking change for some users to remove it. I'm also not sure about the options for IOS users to install custom CA.
Maybe another proposal could be to keep it for now and to improve it later. For example in some applications if you try to reach a server with a self signed cert you have to approve it directly in the app and then this app will remember your choice so if the cert change you'll have to approve it again.

jfly commented on mobile/android/app/src/main/res/xml/network_security_config.xml:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<network-security-config>
+  <base-config>
```
Apparently I was wrong, and my changes here do alter the default. Or at least they did for finamp (see https://github.com/jmshrv/finamp/pull/1276#issuecomment-3130117065).

I've added `cleartextTrafficPermitted="true">` as you originally requested.

shenlong-tanwen commented:
> Concretely, which PRs are we waiting for? Just #21459?

The one just adds the native client deps and uses it for the thumbnails. There will be another PR (yet to be worked on) that migrates the other parts to use the clients as well. We can adapt your changes based on how well they handle the self signed deployments

jfly commented:
@shenlong-tanwen, friendly ping on this PR. With the recent [stable release](https://github.com/immich-app/immich/discussions/22546) of Immich, does that effect how we handle the breaking changes in here?

Also, how is https://github.com/immich-app/immich/pull/21459 looking? Would you be willing to reconsider the ordering here and land my PR first?

richardlt commented:
Hi, any news on that topic ?

denysvitali commented:
FYI: https://github.com/immich-app/immich/pull/22768 supports user-provided certificates and fixes all the problems mentioned in #15230 by using the same client everywhere.

The idea here is to use OkHttp, which supports user certs on Android.

jfly committed:
chore: deprecate the "Allow self-signed SSL certificates" feature

This is unnecessary now that we support user-installed certificates on
Android. Users should be adding certs to their system instead, which
will work in more places (such as browsers), and won't blindly trust the
cert.

shenlong-tanwen reviewed (commented):
Sorry that this got shelved for so long. Can you add the pigeon generator command to the `mise.toml` file in the mobile app directory as well?

The other parts looks good and we can actually merge this sooner now that we have proper backward support for current installs. Apologies again and Thanks a lot for working on this

jfly commented:
I don't care what networking stack we use. OkHttp sounds great if it can also do proper handling of user-installed certificates on Android. I just want to get back to installing Immich from f-droid, rather than infrequently rebasing this branch (and dealing with conflicts).

My preference would be to land this PR as is so certificate handling on android can get un-broken. I'd be happy for @denysvitali to resurrect their work in https://github.com/immich-app/immich/pull/22768 and delete the hack I've added here. I don't think landing this PR increases the effort to switch to OkHttp.

@shenlong-tanwen, WDYT?

mertalev commented:
I totally get that you want to fix this sooner rather than later and I agree! I'm just already working on a more complete solution, so it seems unnecessary to merge this code if it's going to be superseded soon.

jfly commented:
Sorry, I'm getting little frustrated here. This wouldn't be the first time this PR has had to wait for a networking refactor you've done (https://github.com/immich-app/immich/pull/19830#issuecomment-3244229068). This PR exists, is reviewed and ready to land, and fixes a real issue in the wild. It also deprecates a very dangerous feature that IMO shouldn't exist. Why not merge as is? I'll still cheer the day you delete this hack!

mertalev commented:
Sorry for the frustration. This wouldn't be merged until 2.6, by which point I expect the whole app to be switched over to native clients. Merging this PR implies testing it only to revert it right after, which in turn means a larger diff for others in the team to review. It's honestly not a good use of time.

As a team, we have our own planning and this is a task I've already picked up, so please rest assured that I'll take care of it.

richardlt commented:
What a surprise to see this PR closed after all the discussions and changes made to remove any breaking changes.
Hope that we will finally have that fix/feature soon in 2.6 🙏🏼

mertalev commented:
> @mertalev, could you share an issue/PR I can subscribe to so I know how your OkHttp work is going? Thanks, and good luck!

It still needs some work before it's ready for a PR, but I'll ping you when it's up :)

jfly commented:
For anyone else who wants to follow along, @mertalev and I are discussing this over in https://github.com/immich-app/immich/pull/25802
